### ISSUE TITLE
manylinux2014_aarch64/ppc64le/s390x images don't work on Ubuntu 18.04 (but they do on macOS and Ubuntu 20.04)

### ISSUE THREAD

**Issue by ShaneHarvey** [pypa/manylinux#683]
The manylinux2014_aarch64, manylinux2014_ppc64le, and manylinux2014_s390x images work on macOS but not on Linux.

Here you can see that they run on macOS which is x86_64:
```
$ docker version
Client: Docker Engine - Community
 Azure integration  0.1.10
 Version:           19.03.12
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        48a66213fe
 Built:             Mon Jun 22 15:41:33 2020
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.12
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       48a66213fe
  Built:            Mon Jun 22 15:49:27 2020
  OS/Arch:          linux/amd64
  Experimental:     true
 containerd:
  Version:          v1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.15.5
BuildVersion:	19F101

$ uname -a
Darwin Shanes-MacBook-Pro.local 19.5.0 Darwin Kernel Version 19.5.0: Tue May 26 20:41:44 PDT 2020; root:xnu-6153.121.2~2/RELEASE_X86_64 x86_64

$ docker run --rm -it quay.io/pypa/manylinux2014_aarch64 bash
[root@6475fee4244c /]# uname -a
Linux 6475fee4244c 4.19.76-linuxkit #1 SMP Thu Oct 17 19:31:58 UTC 2019 aarch64 aarch64 aarch64 GNU/Linux
[root@6475fee4244c /]# exit

$ docker run --rm -it quay.io/pypa/manylinux2014_ppc64le bash
[root@0448013ebe45 /]# uname -a
Linux 0448013ebe45 4.19.76-linuxkit #1 SMP Thu Oct 17 19:31:58 UTC 2019 ppc64le ppc64le ppc64le GNU/Linux
[root@0448013ebe45 /]# exit

$ docker run --rm -it quay.io/pypa/manylinux2014_s390x bash
bash-4.2# uname -a
Linux de1e669b4e64 4.19.76-linuxkit #1 SMP Thu Oct 17 19:31:58 UTC 2019 s390x s390x s390x GNU/Linux
bash-4.2# exit
```

Here they fail on Linux (Ubuntu 18): 
```
$ docker version
Client: Docker Engine - Community
 Version:           19.03.12
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        48a66213fe
 Built:             Mon Jun 22 15:45:36 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.12
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       48a66213fe
  Built:            Mon Jun 22 15:44:07 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683

$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.2 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.2 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic

$ uname -a
Linux ip-10-122-14-230 4.15.0-1044-aws #46-Ubuntu SMP Thu Jul 4 13:38:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

$ docker run --rm -it quay.io/pypa/manylinux2014_x86_64 bash
Using default tag: latest
latest: Pulling from pypa/manylinux2014_x86_64
524b0c1e57f8: Pull complete
02dfd5116acc: Pull complete
07bdfac9dd6e: Pull complete
Digest: sha256:d5cc9e180ae2c185c892844c4ba995698feb54e106ed6606074554602d9c5cce
Status: Downloaded newer image for quay.io/pypa/manylinux2014_x86_64:latest
quay.io/pypa/manylinux2014_x86_64:latest
[root@d99f16bc71b4 /]# uname -a
Linux d99f16bc71b4 4.15.0-1044-aws #46-Ubuntu SMP Thu Jul 4 13:38:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
[root@d99f16bc71b4 /]# exit

$ docker run --rm -it quay.io/pypa/manylinux2014_aarch64 bash
Using default tag: latest
latest: Pulling from pypa/manylinux2014_aarch64
ae60b79047f8: Pull complete
a00520795f69: Pull complete
abba5409f3f9: Pull complete
Digest: sha256:a17fb2f5d8997718063e28c77a21041cbb65fad10cdb1c879cd38ff7a967eeb0
Status: Downloaded newer image for quay.io/pypa/manylinux2014_aarch64:latest
quay.io/pypa/manylinux2014_aarch64:latest
standard_init_linux.go:211: exec user process caused "no such file or directory"

$ docker run --rm -it quay.io/pypa/manylinux2014_ppc64le bash
Unable to find image 'quay.io/pypa/manylinux2014_ppc64le:latest' locally
latest: Pulling from pypa/manylinux2014_ppc64le
0f49108bec9c: Pull complete
4f5a6a20993c: Pull complete
c822da7239fe: Pull complete
Digest: sha256:b6110c5f4eb0f4a24914935af9ce010d30066ebb0caa3f38b1aa9000514cbc28
Status: Downloaded newer image for quay.io/pypa/manylinux2014_ppc64le:latest
standard_init_linux.go:211: exec user process caused "no such file or directory"

$ docker run --rm -it quay.io/pypa/manylinux2014_s390x bash
Unable to find image 'quay.io/pypa/manylinux2014_s390x:latest' locally
latest: Pulling from pypa/manylinux2014_s390x
0bcda6d6228a: Pull complete
ca0db3dd2e8a: Pull complete
f9f785dc2b3d: Pull complete
Digest: sha256:61daa1b581d357ac8b3fdd08629e128a4796defd829f7626cb7239759e09fbd2
Status: Downloaded newer image for quay.io/pypa/manylinux2014_s390x:latest
standard_init_linux.go:211: exec user process caused "no such file or directory"
```

In https://github.com/pypa/manylinux/issues/410, @hrw says that these docker images need to be run on a host with the same architecture. So supposedly manylinux2014_aarch64 only runs on aarch64 hosts, manylinux2014_ppc64le only runs on ppc64le, and manylinux2014_s390x only runs on s390x. I don't understand that explanation because these images work on macOS which is x86_64. 

Is it possible to run these images on Linux?

**Comment by frenzymadness** [comment]
I am building aarch64/ppc64 wheels on x86_64 Linux machine using qemu-user-static emulation. The emulator exists as a [container image](https://github.com/multiarch/qemu-user-static) or some distributions (like Fedora) have it as an RPM package.

**Comment by jamadden** [comment]
I was suprised (and excited) to see this.

According to [the Docker For Mac documentation](https://docs.docker.com/docker-for-mac/multi-arch/):

> Docker Desktop provides `binfmt_misc` multi-architecture support, which means you can run containers for different Linux architectures such as arm, mips, ppc64le, and even s390x.

> This does not require any special configuration in the container itself as it uses `qemu-static` from the Docker for Mac VM. Because of this, you can run an ARM container, like the arm32v7 or ppc64le variants of the busybox image.

I booted up one of these images and started building one of my projects. Checking the process list, I see that all the processes are being emulated:

```console
$ ps -eaf
UID        PID  PPID  C STIME TTY          TIME CMD
root         1     0  0 10:17 pts/0    00:00:00 /usr/bin/qemu-ppc64le /usr/bin/bash
root        62     1  0 10:19 pts/0    00:00:00 /usr/bin/qemu-ppc64le /usr/bin/bash
root        64    62  1 10:19 pts/0    00:00:04 /usr/bin/qemu-ppc64le /bin/sh ./configure
root     17839    64  0 10:24 pts/0    00:00:00 /usr/bin/qemu-ppc64le /bin/sh ./configure
root     17841 17839  0 10:24 pts/0    00:00:00 /usr/bin/qemu-ppc64le /bin/sh ./configure
```

I don't know what additional pre-configuration Docker/Mac ships with that makes this work out-of-the-box, but I'd guess/hope it should be possible to replicate that same configuration on Linux...

**Comment by ShaneHarvey** [comment]
Aha! Thanks, that explains it! According to [this blog post](https://medium.com/@artur.klauser/building-multi-architecture-docker-images-with-buildx-27d80f7e2408) QEMU and binfmt-support need to be installed on the host system for this to work. All the dependencies are included with Docker Desktop >= 2.1.0 (eg docker for mac/windows).

I haven't been able to get this to work on Ubuntu 18 but I have on Ubuntu 20!
```
$ sudo apt install docker.io
...

$ docker version
Client:
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.13.8
 Git commit:        afacb8b7f0
 Built:             Tue Jun 23 22:26:12 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.8
  Git commit:       afacb8b7f0
  Built:            Thu Jun 18 08:26:54 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.3.3-0ubuntu2
  GitCommit:
 runc:
  Version:          spec: 1.0.1-dev
  GitCommit:
 docker-init:
  Version:          0.18.0
  GitCommit:

$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

$ uname -a
Linux ip-10-122-14-220 5.4.0-1009-aws #9-Ubuntu SMP Sun Apr 12 19:46:01 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

$ docker run --rm -it quay.io/pypa/manylinux2014_aarch64 bash
Using default tag: latest
latest: Pulling from pypa/manylinux2014_aarch64
ae60b79047f8: Pull complete
779a34a2bc80: Pull complete
766f03893317: Pull complete
Digest: sha256:91ae1418c930d6a0259abd44a84b07e0be3755ac8c17aee2bb220b9ea910f30b
Status: Downloaded newer image for quay.io/pypa/manylinux2014_aarch64:latest
quay.io/pypa/manylinux2014_aarch64:latest
[root@7843cc0fb2c5 /]# uname -a
Linux 7843cc0fb2c5 5.4.0-1009-aws #9-Ubuntu SMP Sun Apr 12 19:46:01 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux

$ docker run --rm -it quay.io/pypa/manylinux2014_ppc64le bash
Unable to find image 'quay.io/pypa/manylinux2014_ppc64le:latest' locally
latest: Pulling from pypa/manylinux2014_ppc64le
0f49108bec9c: Pull complete
1360913886fe: Pull complete
24ed96645422: Pull complete
Digest: sha256:e5b7e2462a27e4397b2401b5f996f966ec2abfef34783bf1cb07733df48b5cc5
Status: Downloaded newer image for quay.io/pypa/manylinux2014_ppc64le:latest
[root@c180d4e10393 /]# uname -a
Linux c180d4e10393 5.4.0-1009-aws #9-Ubuntu SMP Sun Apr 12 19:46:01 UTC 2020 ppc64le ppc64le ppc64le GNU/Linux
[root@c180d4e10393 /]# exit

$ docker run --rm -it quay.io/pypa/manylinux2014_s390x bash
Unable to find image 'quay.io/pypa/manylinux2014_s390x:latest' locally
latest: Pulling from pypa/manylinux2014_s390x
0bcda6d6228a: Pull complete
b99381c1b71b: Pull complete
25a4fb413027: Pull complete
Digest: sha256:25a70ac599e84a9510537fc1563ac5c89945cd1ec8cda45a677f6ce30ecc0637
Status: Downloaded newer image for quay.io/pypa/manylinux2014_s390x:latest
bash-4.2# uname -a
Linux 72b4349b0c13 5.4.0-1009-aws #9-Ubuntu SMP Sun Apr 12 19:46:01 UTC 2020 s390x s390x s390x GNU/Linux
```

Feel free to close this issue since I don't think there's anything required by manylinux.

**Comment by mayeut** [comment]
Here's a sample GitHub Actions file demonstrating how to run those images on Ubuntu 18.04:
```
name: QEMU

on: [push, pull_request, workflow_dispatch]

jobs:
  qemu:
    name: QEMU
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
    - name: Pull images
      run: |
        docker pull multiarch/qemu-user-static
        docker pull quay.io/pypa/manylinux2014_aarch64:latest
        docker pull quay.io/pypa/manylinux2014_ppc64le:latest
        docker pull quay.io/pypa/manylinux2014_s390x:latest
    - name: Install QEMU
      run: |
        docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
    - name: Run all images
      run: |
        docker run --rm quay.io/pypa/manylinux2014_aarch64 uname -a
        docker run --rm quay.io/pypa/manylinux2014_ppc64le uname -a
        docker run --rm quay.io/pypa/manylinux2014_s390x uname -a
```

The output of the `Run all images` stage is:
```
Linux 549c89d5535d 5.4.0-1031-azure #32~18.04.1-Ubuntu SMP Tue Oct 6 10:03:22 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux
Linux 8d086191af67 5.4.0-1031-azure #32~18.04.1-Ubuntu SMP Tue Oct 6 10:03:22 UTC 2020 ppc64le ppc64le ppc64le GNU/Linux
Linux b26e4043047e 5.4.0-1031-azure #32~18.04.1-Ubuntu SMP Tue Oct 6 10:03:22 UTC 2020 s390x s390x s390x GNU/Linux
```

https://github.com/mayeut/pep600_compliance/runs/1338868363?check_suite_focus=true
